File: typescript/src/graph/bipartite-matching.ts
// LombokAlgoritma — Maximum bipartite matching (Hopcroft–Karp)
// SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok

import { OutOfRangeError } from '../core/errors.js';
/** Result of {@link bipartiteMatching}. */
export interface MatchingResult {
  /** Number of matched pairs (normative, SPEC §6.14). */
  size: number;
  /** `matchLeft[u]` = matched right vertex of left vertex `u`, or −1. */
  matchLeft: number[];
  /** `matchRight[v]` = matched left vertex of right vertex `v`, or −1. */
  matchRight: number[];
}

/**
 * Maximum matching of the bipartite graph with left vertices `0 … nLeft − 1`, right vertices
 * `0 … nRight − 1` and `pairs` = `[left, right]` edges. Hopcroft–Karp, O(E·√V).
 * The matching *size* is normative; which maximum matching is returned is not.
 *
 * @throws RangeError for an edge outside the vertex ranges
 */
export function bipartiteMatching(
  nLeft: number,
  nRight: number,
  pairs: ReadonlyArray<readonly [number, number]>,
): MatchingResult {
  const adj: number[][] = Array.from({ length: nLeft }, () => []);
  for (const [u, v] of pairs) {
    if (
      !(Number.isInteger(u) && Number.isInteger(v) && u >= 0 && u < nLeft && v >= 0 && v < nRight)
    ) {
      throw new OutOfRangeError(`bipartiteMatching: edge [${u}, ${v}] out of range`);
    }
    (adj[u] as number[]).push(v);
  }
  const matchLeft = new Array<number>(nLeft).fill(-1);
  const matchRight = new Array<number>(nRight).fill(-1);
  const dist = new Array<number>(nLeft).fill(0);
  const INF = Number.POSITIVE_INFINITY;

  const bfs = (): boolean => {
    const q: number[] = [];
    for (let u = 0; u < nLeft; u++) {
      if (matchLeft[u] === -1) {
        dist[u] = 0;
        q.push(u);
      } else dist[u] = INF;
    }
    let found = false;
    for (let h = 0; h < q.length; h++) {
      const u = q[h] as number;
      for (const v of adj[u] ?? []) {
        const w = matchRight[v] as number;
        if (w === -1) found = true;
        else if (dist[w] === INF) {
          dist[w] = (dist[u] as number) + 1;
          q.push(w);
        }
      }
    }
    return found;
  };

  // iterative DFS along the layered graph
  const it = new Array<number>(nLeft).fill(0);
  const tryAugment = (root: number): boolean => {
    const stack = [root];
    const via: number[] = []; // right vertex chosen at each level
    while (stack.length > 0) {
      const u = stack[stack.length - 1] as number;
      const nb = adj[u] ?? [];
      let pushed = false;
      while ((it[u] as number) < nb.length) {
        const v = nb[it[u] as number] as number;
        it[u] = (it[u] as number) + 1;
        const w = matchRight[v] as number;
        if (w === -1) {
          via.push(v);
          for (let k = stack.length - 1; k >= 0; k--) {
            const lu = stack[k] as number;
            const rv = via[k] as number;
            matchLeft[lu] = rv;
            matchRight[rv] = lu;
          }
          return true;
        }
        if (dist[w] === (dist[u] as number) + 1) {
          via.push(v);
          stack.push(w);
          pushed = true;
          break;
        }
      }
      if (pushed) continue;
      dist[u] = INF;
      stack.pop();
      via.pop();
    }
    return false;
  };

  let size = 0;
  while (bfs()) {
    it.fill(0);
    for (let u = 0; u < nLeft; u++) if (matchLeft[u] === -1 && tryAugment(u)) size++;
  }
  return { size, matchLeft, matchRight };
}

File: python/lombokalgoritma/graph/bipartite_matching.py
# LombokAlgoritma — maximum bipartite matching, Hopcroft–Karp (SPEC §9.12)
# SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok
from __future__ import annotations

import math
from collections.abc import Sequence
from typing import NamedTuple

from ..errors import OutOfRangeError


class MatchingResult(NamedTuple):
    """Matching size (normative) and the partner of every vertex (−1 when unmatched)."""

    size: int
    match_left: list[int]
    match_right: list[int]


def _is_int(x: object) -> bool:
    return isinstance(x, int) and not isinstance(x, bool)


def bipartite_matching(
    n_left: int, n_right: int, pairs: Sequence[tuple[int, int] | Sequence[int]]
) -> MatchingResult:
    """Maximum matching of left ``0 … n_left − 1`` / right ``0 … n_right − 1``. O(E·√V).

    Raises:
        OutOfRangeError: a pair outside the vertex ranges.
    """
    adj: list[list[int]] = [[] for _ in range(n_left)]
    for pr in pairs:
        u, v = pr[0], pr[1]
        if not (_is_int(u) and _is_int(v) and 0 <= u < n_left and 0 <= v < n_right):
            raise OutOfRangeError(f"bipartite_matching: edge [{u}, {v}] out of range")
        adj[u].append(v)
    match_left = [-1] * n_left
    match_right = [-1] * n_right
    inf = math.inf
    dist: list[float] = [0] * n_left

    def bfs() -> bool:
        q: list[int] = []
        for u in range(n_left):
            if match_left[u] == -1:
                dist[u] = 0
                q.append(u)
            else:
                dist[u] = inf
        found = False
        h = 0
        while h < len(q):
            u = q[h]
            h += 1
            for v in adj[u]:
                w = match_right[v]
                if w == -1:
                    found = True
                elif dist[w] == inf:
                    dist[w] = dist[u] + 1
                    q.append(w)
        return found

    it = [0] * n_left

    def try_augment(root: int) -> bool:
        stack = [root]
        via: list[int] = []
        while stack:
            u = stack[-1]
            nb = adj[u]
            pushed = False
            while it[u] < len(nb):
                v = nb[it[u]]
                it[u] += 1
                w = match_right[v]
                if w == -1:
                    via.append(v)
                    for lu, rv in zip(stack, via, strict=True):
                        match_left[lu] = rv
                        match_right[rv] = lu
                    return True
                if dist[w] == dist[u] + 1:
                    via.append(v)
                    stack.append(w)
                    pushed = True
                    break
            if pushed:
                continue
            dist[u] = inf
            stack.pop()
            if via:
                via.pop()
        return False

    size = 0
    while bfs():
        for i in range(n_left):
            it[i] = 0
        for u in range(n_left):
            if match_left[u] == -1 and try_augment(u):
                size += 1
    return MatchingResult(size, match_left, match_right)
